src/Microsoft.Azure.WebJobs.Extensions.Redis/StreamTrigger/RedisStreamTriggerScaleMonitor.cs

```diff
@@ -11,21 +12,71 @@ internal class RedisStreamTriggerScaleMonitor : RedisPollingTriggerBaseScaleMoni
-            : base(multiplexer, maxBatchSize, key)
+            : base(name, multiplexer, maxBatchSize, key)
         {
-            this.Descriptor = new ScaleMonitorDescriptor(name, $"{name}-RedisStreamTrigger-{key}");
-            this.TargetScalerDescriptor = new TargetScalerDescriptor($"{name}-RedisStreamTrigger-{key}");
+            this.Descriptor = new ScaleMonitorDescriptor(name, RedisScalerProvider.GetFunctionScalerId(name, "RedisStreamTrigger", key));
+            this.TargetScalerDescriptor = new TargetScalerDescriptor(RedisScalerProvider.GetFunctionScalerId(name, "RedisStreamTrigger", key));
         }
 
-        public override Task<RedisPollingTriggerBaseMetrics> GetMetricsAsync()
+        public override async Task<RedisPollingTriggerBaseMetrics> GetMetricsAsync()
         {
-            var metrics = new RedisPollingTriggerBaseMetrics
+            long streamLength = await multiplexer.GetDatabase().StreamLengthAsync(key);
+            if (streamLength == 0)
             {
-                Remaining = multiplexer.GetDatabase().StreamLength(key),
+                return new RedisPollingTriggerBaseMetrics
+                {
+                    Remaining = 0,
+                    Timestamp = DateTime.UtcNow,
+                };
+            }
+
+            StreamGroupInfo[] groups = multiplexer.GetDatabase().StreamGroupInfo(key);
+            if (!groups.Any(g => g.Name == name))
+            {
+                // group doesn't exist, assume all entries in stream have not been processed
+                return new RedisPollingTriggerBaseMetrics
+                {
+                    Remaining = streamLength,
+                    Timestamp = DateTime.UtcNow,
+                };
+            }
+
+            StreamGroupInfo group = groups.First(g => g.Name == name);
+            if (multiplexer.GetServers()[0].Version >= RedisUtilities.Version70 && group.Lag.HasValue)
+            {
+                // Redis 7: Scaler gets number of remaining entries for the consumer group from XINFO GROUPS.
+                return new RedisPollingTriggerBaseMetrics
+                {
+                    Remaining = group.Lag.Value,
+                    Timestamp = DateTime.UtcNow,
+                };
+            }
+
+            /*
+             * Redis 6/6.2 does not have an internal counter for the number of remaining entries for the consumer group to read
+             * Estimate number of remaining entries in the stream using the unixtime field from the entry IDs:
+             * - default ID structure ("unixtime-counter")
+             * - entries are added at a constant rate over time
+            */
+            StreamInfo info = await multiplexer.GetDatabase().StreamInfoAsync(key);
+            string firstId = (string)info.FirstEntry.Id;
+            string lastId = (string)info.LastEntry.Id;
+            string lastDeliveredId = group.LastDeliveredId ?? firstId;
+
+            long firstTimestamp = long.Parse(firstId.Split('-')[0]);
```

Comment:
If the entity id is not the default format, and we get it wrong, what will be the worst consequence?

Reply:
I've updated the logic to ensure that as long as there are any unprocessed entries in the streams, the value returned will be between 1 and the stream length. If all entries are processed, then it will return 0.
Even if the ID is custom, the ID format `<number>-<counter>` is the same, and the IDs must strictly be increasing as noted [here](https://redis.io/commands/xadd). Even with custom IDs, the estimation logic will give somewhere within the range.